src/main/scala/is/hail/utils/UnionFind.scala

```diff
@@ -0,0 +1,62 @@
+package is.hail.utils
+
+class UnionFind(initialCapacity: Int = 32) {
+  private var a: Array[Int] = new Array[Int](initialCapacity)
+  private var rank: Array[Int] = new Array[Int](initialCapacity)
+  private var count: Int = 0
+
+  def size: Int = count
+
+  private def ensure(i: Int) {
+    if (i >= a.length) {
+      var newLength = a.length
+      do {
+        newLength = newLength << 1
+      } while (i >= newLength);
+      val a2 = new Array[Int](newLength)
+      Array.copy(a, 0, a2, 0, a.length)
+      a = a2
+      val rank2 = new Array[Int](newLength)
+      Array.copy(rank, 0, rank2, 0, rank.length)
+      rank = rank2
+    }
+  }
+
+  def makeSet(i: Int) {
+    ensure(i)
+    a(i) = i
+    count += 1
+  }
+
+  def find(x: Int): Int = {
+    var representative = x
+    while (representative != a(representative)) {
+      representative = a(representative)
+    }
+    var current = x
+    while (representative != current) {
+      val temp = a(current)
+      a(current) = representative
```

Comment:
I presume this is to remove intermediate nodes to the representative node. Otherwise, you could just return after the first while loop.

Reply:
Yes exactly. Wikipedia describes this as "path compression" and it is necessary to get the most restrictive asymptotic bounds on find and union (finds tend to be constant time since if your or any descendants get "found" you'll get your parent set to the representative)